pyroute2/netlink/rtnl/rtmsg.py

```diff
@@ -160,7 +160,144 @@ class seg6_encap_info(nla):
+
+            # Corresponding values for seg6 encap modes
+            SEG6_IPTUN_MODE_INLINE = 0
+            SEG6_IPTUN_MODE_ENCAP = 1
+
+            # Mapping string to nla value
+            encapmodes = {
+              "inline": SEG6_IPTUN_MODE_INLINE,
+              "encap": SEG6_IPTUN_MODE_ENCAP
+            }
+
+            # Nla value for seg6 type
+            SEG6_TYPE = 4
+
+            # Flag value for hmac
+            SR6_FLAG1_HMAC = 1 << 3
+
+            # Tlv value for hmac
+            SR6_TLV_HMAC = 5
+
+            # Utility function to get the family from the msg
+            def get_family(self):
+                pointer = self
+                while pointer.parent is not None:
+                    pointer = pointer.parent
+                return pointer.get('family', AF_UNSPEC)
+
+            def encode(self):
+                # Retrieve the family
+                family = self.get_family()
+                # Seg6 can be applied only to IPv6
+                if family == AF_INET6:
+                    # Get mode
+                    mode = self['mode']
+                    # Get segs
+                    segs = self['segs']
+                    # Get hmac
+                    hmac = self.get('hmac', None)
+                    # With "inline" mode there is not
+                    # encap into an outer IPv6 header
+                    if mode == "inline":
+                        # Add :: to segs
+                        segs.insert(0, "::")
+                    # Add mode to value
+                    self['encapmode'] = self.encapmodes.get(mode, self.SEG6_IPTUN_MODE_ENCAP)
+                    # Calculate srlen
+                    srhlen = 8 + 16 * len(segs);
+                    # If we are using hmac we have a tlv as trailer data
+                    if hmac:
+                        # Since we can use sha1 or sha256
+                        srhlen += 40
+                    # Calculate and set hdrlen
+                    self['hdrlen'] = (srhlen >> 3) - 1
+                    # Add seg6 type
+                    self['type'] = self.SEG6_TYPE
+                    # Add segments left
+                    self['segments_left'] = len(segs) - 1
+                    # Add fitst segment
+                    self['first_segment'] = len(segs) - 1
+                    # If hmac is used we have to set the flags
+                    if hmac:
+                        # Add SR6_FLAG1_HMAC
+                        self['flags'] |= self.SR6_FLAG1_HMAC
+                    # Init segs
+                    self['segs'] = b''
+                    # Iterate over segments
+                    for seg in segs:
+                        # Convert to network byte order and add to value
+                        self['segs'] += inet_pton(family, seg)
+                    # Initialize tlvs
+                    self['tlvs'] = b''
+                    # If hmac is used we have to properly init tlvs
+                    if hmac:
```

Comment:
Maybe this part can be optimized. Any idea ?

Reply:
Any string concatenation via `+` is not optimal in Python, but we can optimize that later.

The most important is that the code is readable. It is much more important even than the performance, from my experience.